_posts/2021-12-08-kafkasql-storage-and-security.md

```diff
@@ -0,0 +1,471 @@
+
+*Note: If you have not created the topic manually, Apicurio Registry will create the `kafkasql-journal` automatically.
+Make sure you create a CDR for the `kafkasql-journal` topic, so you can manage it with Strimzi.*
+
+# Reproducer
+
+Let's create a procedure to attempt to reproduce the issue and confirm our hypothesis. 
+We are assuming that the Kafka cluster and the Apicurio Registry executable is already prepared by following steps in the previous sections.
+
+1. Stop the Registry process.
+1. Remove `kafkasql-journal` topic, if it exists.
+1. Create the topic manually with aggressive log compaction settings:
+   
+    ```shell
+    bin/kafka-topics.sh --bootstrap-server localhost:9092 \
+       --create --topic kafkasql-journal --partitions 1 --replication-factor 1 \
+       --config min.cleanable.dirty.ratio=0.000001 --config cleanup.policy=compact \
+       --config segment.ms=100 --config delete.retention.ms=100
+    ```
+1. Create at least 3 AVRO schemas, for example:
+   
+    ```json
+    {
+     "type": "record",
+     "name": "User1",
+     "namespace": "foo",
+     "fields": [
+       { "name": "active1", "type": ["null", "boolean"] }
+     ]
+    }
+    ```
+    *Sample Avro schema*
+
+1. Restart the registry, so it will reload messages from the topic into its local database. 
+   Remove the database file if you are using the debug settings from earlier. 
+
+1. Add another AVRO schema to the Registry and observe the error.
+
+# Solution
+
+In order to fix the bug, we have to prevent messages, which control the database sequence, from being compacted. 
+We are already doing this for other types of messages, by making sure that the key value is always unique. 
+We can do this by including a sufficiently random nonce, such as an `UUID`, in the message key.
+
+The following is an example `ContentId` message, that increments a sequence for the `CONTENT` table, before the fix has been implemented:
+
+```json
+{
+  "topic": "kafkasql-journal",
+  "partition": 0,
+  "offset": 1,
+  "tstype": "create",
+  "ts": 1638894723320,
+  "headers": [
+    "req",
+    "9dff3fb8-7a1d-41b0-9083-684d9fbfa619"
+  ],
+  "key": "\t{\"tenantId\":\"_\"}",
+  "payload": "\t{\"action\":\"CREATE\"}"
+}
+```
+
+And the same message after the bugfix:
+
+```json
+{
+  "topic": "kafkasql-journal",
+  "partition": 0,
+  "offset": 1,
+  "tstype": "create",
+  "ts": 1638894723320,
+  "headers": [
+    "req",
+    "9dff3fb8-7a1d-41b0-9083-684d9fbfa619"
+  ],
+  "key": "\t{\"tenantId\":\"_\",\"uuid\":\"6875bbc9-92fe-4572-accc-7dab2bdcd84e\"}",
+  "payload": "\t{\"action\":\"CREATE\"}"
+}
+```
+
+# Temporary Resolution for Older Apicurio Registry Versions
+
+This issue has already been fixed in Apicurio Registry `2.1.4.Final`. If you are not able to upgrade, follow this procedure:
+
+In order to make sure you are not affected by this bug, you need to disable log compaction for the `kafkasql-journal` topic
+in your cluster. First, check if it is enabled:
```

Comment:
Suggest adding something like `However, if you disable log compaction you MUST make sure to also properly set message persistence to 'forever' so that messages are not lost from the topic`